### PR TITLE
Lookup methods in constructors

### PR DESCRIPTION
**Pull request by ricea**
Lookup methods on underlyingSource, underlyingSink and transformer in
the constructor. The main benefit is that identity transforms can be
simply implemented without touching Javascript. Other less critical
optimisation opportunities are unlocked for other stream types.

The algorithm construction for used-supplied methods is changed so that
a different algorithm is constructed depending on whether or not the
method was supplied. Two new abstract operations are added:

* GetMethod works like GetV but throws if the value is neither undefined
  nor a function.
* PromiseInvoke works like PromiseInvokeOrNoop but doesn't check for
  undefined.

PromiseInvokeOrNoop is no longer used and has been removed.

Constructors now throw for invalid method parameters, which results in
some test expectation changes. Changing the methods or prototype on the
underlying object after construction no longer does anything. Other than
these edge cases there are no web-developer visible changes.

Fixes https://github.com/whatwg/streams/issues/691.

***
<a href="https://whatpr.org/streams/860.html" title="Last updated on Dec 6, 2017, 9:32 AM GMT (b90b687)">Preview</a> | <a href="https://whatpr.org/streams/860/5b9f39d...b90b687.html" title="Last updated on Dec 6, 2017, 9:32 AM GMT (b90b687)">Diff</a>

### CODE REVIEW

**Comment by ricea** [comment]
Only the reference implementation changes are done so far.

The algorithm construction is very repetitive. Part of the problem is that the controller is not available at the time we construct the algorithms. Suggestions welcome.

File: reference-implementation/lib/helpers.js [hunk 1]
```diff
@@ -83,15 +83,20 @@ exports.InvokeOrNoop = (O, P, args) => {
   return Call(method, O, args);
 };
 
-exports.PromiseInvokeOrNoop = (O, P, args) => {
-  assert(O !== undefined);
-  assert(IsPropertyKey(P));
-  assert(Array.isArray(args));
+exports.PromiseInvoke = (F, V, args) => {
```
Comment (domenic) [inline reference-implementation/lib/helpers.js]: This is no longer an "Invoke", but instead a "Call". "Invokes" take (object, property name) pairs.

File: reference-implementation/lib/readable-stream.js [hunk 2]
```diff
@@ -1156,12 +1156,22 @@ function SetUpReadableStreamDefaultControllerFromUnderlyingSource(stream, underl
     return InvokeOrNoop(underlyingSource, 'start', [stream._readableStreamController]);
   }
 
-  function pullAlgorithm() {
-    return PromiseInvokeOrNoop(underlyingSource, 'pull', [stream._readableStreamController]);
+  // eslint-disable-next-line func-style
```
Comment (domenic) [inline reference-implementation/lib/readable-stream.js]: I see what you mean about the controller not existing yet being a problem for algorithm construction. Can we just move the single line

```js
const controller = Object.create(ReadableStreamDefaultController.prototype);
```

before the algorithm creation? Then perhaps pass that as a parameter to the SetUpXYZ() abstract ops? I guess maybe also move the line

```js
stream._readableStreamController = controller;
```

and then we can pass in `controller` instead of `stream` to the SetUpXYZ() ops.

File: reference-implementation/lib/helpers.js [hunk 3]
```diff
@@ -83,15 +83,20 @@ exports.InvokeOrNoop = (O, P, args) => {
...
-    return Promise.resolve(exports.InvokeOrNoop(O, P, args));
-  } catch (returnValueE) {
-    return Promise.reject(returnValueE);
+    return Promise.resolve(Call(F, V, args));
+  } catch (value) {
+    return Promise.reject(value);
+  }
+};
+
+exports.GetMethod = (V, methodName) => {
```
Comment (domenic) [inline reference-implementation/lib/helpers.js]: This might end up inlined into whatever helper is used for deduping the algorithm creation code.

File: reference-implementation/lib/readable-stream.js [hunk 4]
```diff
@@ -1156,12 +1156,22 @@ function SetUpReadableStreamDefaultControllerFromUnderlyingSource(stream, underl
...
+  const pullMethod = GetMethod(underlyingSource, 'pull');
+  if (pullMethod !== undefined) {
+    pullAlgorithm = () => {
+      return PromiseInvoke(pullMethod, underlyingSource, [stream._readableStreamController]);
+    };
   }
 
-  function cancelAlgorithm(reason) {
-    return PromiseInvokeOrNoop(underlyingSource, 'cancel', [reason]);
+  // eslint-disable-next-line func-style
```
Comment (domenic) [inline reference-implementation/lib/readable-stream.js]: We should probably just disable this lint in the eslintrc, given that it's hurting us repeatedly in this way. I don't like one-off lint disables like this in general.

File: reference-implementation/lib/transform-stream.js [hunk 5]
```diff
@@ -259,30 +259,44 @@ function SetUpTransformStreamDefaultController(stream, transformAlgorithm, flush
 }
 
 function SetUpTransformStreamDefaultControllerFromTransformer(stream, transformer) {
-  function transformAlgorithm(chunk) {
```
Comment (domenic) [inline reference-implementation/lib/transform-stream.js]: That got a lot simpler, yay.

File: reference-implementation/lib/helpers.js (hunk 1 above)
Comment (ricea) [reply reference-implementation/lib/helpers.js]: Thanks, fixed.

File: reference-implementation/lib/helpers.js (hunk 3 above)
Comment (ricea) [reply reference-implementation/lib/helpers.js]: It turns out TransformStream still needs it to look up the "transform" method.

File: reference-implementation/lib/readable-stream.js (hunk 2 above)
Comment (ricea) [reply reference-implementation/lib/readable-stream.js]: This worked, thanks. I chose to pass both stream and controller and still leave tying them together to SetUpXYZ().

File: reference-implementation/lib/readable-stream.js (hunk 4 above)
Comment (ricea) [reply reference-implementation/lib/readable-stream.js]: Done.

File: reference-implementation/lib/transform-stream.js (hunk 5 above)
Comment (ricea) [reply reference-implementation/lib/transform-stream.js]: Not as simple as I'd like, but all the other algorithm creation steps are looking great now.

**Comment by ricea** [comment]
There's still a linking error for GetV. Apart from that, I think it's pretty much done.

**Comment by ricea** [comment]
I think this is ready to merge.

The wording "Let _fullArgs_ be a List consisting of _arg_ followed by the elements of _extraArgs_ in order." could possibly be improved. I couldn't find an appropriate operation defined in the ecmascript standard and I didn't want to write it out in terms of low-level element manipulations.

**Comment by domenic** [comment]
Oh, hmmm. I realized we have our own GetMethod which is now different from the ES spec's. We should remove it, and I can do that from the spec side, but this is a slight normative change in that it allows `null` as equivalent to undefined. I don't really like that change, but it's probably better to be consistent with ES than to not be. We'll need to update the tests a bit.

**Comment by domenic** [comment]
I wonder if we should also be consistent and allow `size` to be null? Gah, this kind of sucks... What do you think we should do?

**Comment by domenic** [comment]
I see a few options:

- Back out my GetMethod changes, pretend for a few days that we didn't notice this problem, and solve it ASAP.
- Back out my GetMethod changes, rename our GetMethod to GetMethod2, open an issue to fix it, add a note in the spec pointing to that issue, and then merge this PR using GetMethod2
- Keep my GetMethod changes, and fix everything in this PR, probably including `size`. (The way I'd do that is by no longer doing the destructuring for the strategies, then maybe creating a mode for CreateAlgorithmFromUnderlyingMethod or a second abstract op, which can be used for both `start` and `size`.)

WDYT?

**Comment by ricea** [comment]
I would be happier if I knew why the ES version of GetMethod squashes `null` to `undefined`. I can't make sense of it.
```javascript
new TransformStream({
  transform: null
});
```
doesn't look like a sensible way to create an identity transform, it just looks like nonsense to me. It's not like we intend for people to write `{ transform: undefined }`, there's just no point in disallowing it.

So, in the absence of a compelling justification for squashing `null` to `undefined`, I'd prefer to keep our existing semantics. I plan to proceed with option 4: inline GetMethod as it's only used in two places anyway.
